File: src/main/kotlin/de/wittenbude/exportify/models/SpotifyUser.kt
@file:Suppress("PLUGIN_IS_NOT_ENABLED")

package de.wittenbude.exportify.models

import com.neovisionaries.i18n.CountryCode
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import se.michaelthelin.spotify.enums.ModelObjectType
import se.michaelthelin.spotify.enums.ProductType
import se.michaelthelin.spotify.model_objects.specification.User

@Serializable
data class SpotifyUser(
    val birthdate: String?,
    val country: CountryCode?,
    val displayName: String,
    val email: String?,
    val externalUrls: Map<String, String>?,
    val followerHref: String?,
    val followerTotal: Int?,
    val href: String?,
    @SerialName("_id") val id: String,
    val images: List<String>?,
    val product: ProductType?,
    val type: ModelObjectType?,
    val uri: String?
)

fun User.convert(): SpotifyUser {
    return SpotifyUser(
        birthdate,
        country,
        displayName,
        email,
        externalUrls?.externalUrls,
        followers?.href,
        followers?.total,
        href,
        id,
        images?.map { it.url },
        product,
        type,
        uri
    )
}

File: pkg/storage/models/spotifyuser.go
package models

import (
	"expandify-api/pkg/expandify"
	"expandify-api/pkg/storage/models/embeds"
	"gorm.io/datatypes"
)

type SpotifyUser struct {
	DisplayName  string
	ExternalURLs datatypes.JSONType[map[string]string] `json:"external_urls"`
	Followers    embeds.Followers                      `json:"followers" gorm:"embedded;embeddedPrefix:followers_"`
	Href         string                                `json:"href"`
	ID           string                                `json:"id" gorm:"primarykey"`
	Images       datatypes.JSONType[[]embeds.Image]    `json:"images"`
	URI          string                                `json:"uri"`
	Country      string                                `json:"country"`
	Email        string                                `json:"email"`
	Product      string                                `json:"product"`
	Birthdate    string                                `json:"birthdate"`
}

func NewSpotifyUser(spotifyUser *expandify.SpotifyUser) *SpotifyUser {
	var images []embeds.Image
	for _, image := range spotifyUser.Images {
		images = append(images, embeds.NewImage(&image))
	}

	return &SpotifyUser{
		DisplayName:  spotifyUser.DisplayName,
		ExternalURLs: datatypes.NewJSONType(spotifyUser.ExternalURLs),
		Followers:    embeds.NewFollowers(&spotifyUser.Followers),
		Href:         spotifyUser.Href,
		ID:           spotifyUser.SpotifyID,
		Images:       datatypes.NewJSONType(images),
		URI:          spotifyUser.URI,
		Country:      spotifyUser.Country,
		Email:        spotifyUser.Email,
		Product:      spotifyUser.Product,
		Birthdate:    spotifyUser.Birthdate,
	}
}

func (s *SpotifyUser) Convert() *expandify.SpotifyUser {
	var images []expandify.Image
	for _, image := range s.Images.Data() {
		images = append(images, image.Convert())
	}

	return &expandify.SpotifyUser{
		DisplayName:  s.DisplayName,
		ExternalURLs: s.ExternalURLs.Data(),
		Followers:    s.Followers.Convert(),
		Href:         s.Href,
		SpotifyID:    s.ID,
		Images:       images,
		URI:          s.URI,
		Country:      s.Country,
		Email:        s.Email,
		Product:      s.Product,
		Birthdate:    s.Birthdate,
	}
}
